M1FExplained/Chapter25/Exercises/Exercise07.lean

```diff
@@ -0,0 +1,351 @@
+
+
+
+instance : Mul set1 where
+  mul a b := ⟨a.1*b.1,closed_mul a b⟩
+
+@[simp,norm_cast]  
+lemma set1_coe_mul (a b : set1) : (a * b : set1) = (a : ℂ) * b := by rfl
+
+lemma closed_mul2 (a b : set2) : (a : ℤ) * b ∈ set2 := by
+  cases' a with a ha
+  cases' b with b hb
+  rcases ha with (rfl|rfl) <;> rcases hb with (rfl|rfl) <;> norm_num
+
+instance : Mul set2 where
+  mul a b := ⟨a*b,closed_mul2 a b⟩
+
+@[norm_cast]
+lemma set2_coe_mul (a b : set2): (a * b : set2) = (a : ℤ) * b := by rfl
+
+lemma closed_pow (a : set1) (c : set2) : (a : ℂ) ^ (c : ℤ) ∈ set1 := by
+  cases' a with a ha
+  cases' c with b hc
+  rcases ha with (rfl|rfl|rfl|rfl) <;> rcases hc with (rfl|rfl) <;> norm_num
+  right
+  right
+  left
+  simp
+  rw [inv_neg]
+  norm_num
+
+
+noncomputable instance : HPow set1 set2 set1 where
+  hPow a c := ⟨a.1^c.1, closed_pow a c⟩   
+
+@[simp,norm_cast]
+lemma coe_pow (a : set1) (b : set2) : (a^b : set1) = (a : ℂ)^(b : ℤ) := by rfl
+
+noncomputable instance : Mul D where 
+  mul a b := ⟨a.1*b.1^a.2,a.2*b.2⟩
+
+@[simp]
+lemma mul_first (a b : D): (a * b).fst = a.1*b.1^a.2 := by rfl
+
+@[simp]
+lemma mul_snd (a b : D) : (a * b).snd = a.2 * b.2 := by rfl
+
+lemma mul_assocD (a b c : D): a*b*c = a * (b * c) := by
+  unfold D
+  ext1
+  · repeat rw[mul_first]
+    rw [mul_snd]
+    ext1
+    push_cast
+    rw [mul_zpow,mul_comm (a.snd : ℤ),zpow_mul,mul_assoc]
+  · repeat rw [mul_snd]
+    ext
+    push_cast
+    rw [mul_assoc]
+
+instance : One D := ⟨(1,1)⟩ 
+
+@[simp]
+lemma one_first : (1 : D).fst = 1 := by rfl
+
+@[simp,norm_cast]
+lemma set_two_coe_one : ((1 : set2) : ℤ) = 1 := rfl
+
+
+lemma one_mulD (a : D) : 1*a = a := by
+  unfold D
+  ext1
+  · rw [mul_first]
+    ext1
+    push_cast
+    simp
+  · rw [mul_snd]
+    ext
+    push_cast
+    simp
+
+lemma mul_oneD (a : D) : a*1 = a := by
+  unfold D
+  ext1
+  · rw [mul_first]
+    ext1
+    push_cast
+    simp
+  · rw [mul_snd]
+    ext
+    push_cast
+    simp
+
+lemma neg_set2 (a : set2) : -a.1 ∈ set2 := by
+  cases' a with a ha
+  rcases ha with (rfl|rfl)  <;> norm_num
+
+instance : Neg set2 where 
+  neg a := ⟨-a.1, neg_set2 a⟩  
+
+@[simp,norm_cast]
+lemma neg_coe (a : set2) : (-a : set2) = -(a : ℤ) := by rfl
+
+
+noncomputable instance : Inv D where
+  inv a := ⟨a.1^(-a.2),a.2⟩
+
+
+@[simp]
+lemma inv_first (a : D): (a⁻¹).fst = a.1^(-a.2) := by rfl
+
+@[simp]
+lemma inv_snd (a : D) : (a⁻¹).snd = (a.2) := by rfl
+
+lemma set1_ne_zero (a : set1) : (a : ℂ) ≠ 0 := by
+  cases' a with a ha
+  rcases ha with (rfl|rfl|rfl|rfl)  <;> norm_num
+  intro h
+  apply_fun Complex.im at h
+  simp at h
+  intro h
+  apply_fun Complex.im at h
+  simp at h
+
+lemma set.mul_self (a : set2) : (a : ℤ) * a = 1 := by
+  cases' a with a ha
+  rcases ha with (rfl|rfl)  <;> norm_num
+
+lemma mul_left_invD (a : D): a⁻¹ * a = 1 := by
+  unfold D
+  ext1
+  · simp
+    rw [mul_first]
+    ext1
+    push_cast
+    rw [inv_first, inv_snd]
+    push_cast
+    simp
```

Comment:
This is either a non-terminal simp (if you have one goal) or you should be using {} brackets (if you have more than one goal)